### ISSUE TITLE
Merging ReQue and RePair

### ISSUE THREAD

**Issue by DelaramRajaei** [fani-lab/RePair#42]
This is the issue where I log all my processes while adding ReQue's expanders to RePair.

**Comment by hosseinfani** [comment]
@DelaramRajaei Awesome! Thanks.

**Comment by DelaramRajaei** [comment]
Hey @hosseinfani , 

I wanted to provide you with a project update. Currently, the pipeline is operational, although I'm addressing some minor bugs related to reading different datasets. I've initiated backtranslation on two datasets, robust04 and dbpedia, across 10 languages. Below are the logs.
[robust04_dbpedia_backtranslation.zip](https://github.com/fani-lab/RePair/files/13442180/robust04_dbpedia_backtranslation.zip)

**Comment by DelaramRajaei** [comment]
Hey @hosseinfani,
 
I wanted to give you an update on the project.

The successful merger of ReQue and RePair is now complete. I have executed backtranslation for all five datasets, employing two IR rankers (BM25, QLD) and two evaluation metrics (MAP, MRR).

Encountered challenges in loading different datasets, particularly with clueweb09b and gov2, which have split their queries across multiple trecs. Currently, the code reads all files, but I plan to modify it to run each trec separately and aggregate the results, following the approach used in the ReQue project.

Presently, the project is running for all expanders for gov2 across various IR rankers and evaluation metrics. The log of the ongoing run is provided.

[logs.zip](https://github.com/fani-lab/RePair/files/13604926/logs.zip)

The log file contains records for Backtranslation, Conceptnet, Thesaurus, Wordnet, and Tagme refiners. I have also updated the RePair_StoryBoard in the Query Refinement channel on Teams.

In parallel, I am working on the query class and rag fusion, though there hasn't been significant success in those areas yet. I am ensuring the expanders run flawlessly and addressing other bugs.

Additionally, a minor change has been made in the output structure. After creating a folder for each dataset, it will store the refined data there and subsequently store the results of the ranker and metric in a new folder within the dataset folder. Below is an overview of the file storage:

```
├── output
│   ├── gov2 [Dataset's name]
│   │   ├── refined_queries_files
│   │   └── ranker.metric [such as bm25.map]
│   │       └── [This is where all the results from the search, eval, aggregate, and boxing are stored]
```

**Comment by hosseinfani** [comment]
Hi @DelaramRajaei 
Thanks for the update. This is great.
We need a meeting to demo a sample run for me.

**Comment by DelaramRajaei** [comment]
Hello, @hosseinfani 

I am currently facing issues with the `RelevanceFeedback` refiner. As we have transitioned from using Anserini to only Pyserini, one potential solution involves utilizing `SimpleSearcher` from Pyserini. However, this approach encounters problems with multiprocessing (`multiprocessing as mp`), which is deprecated, and the library suggests using `Lucene`. Unfortunately, I couldn't find a similar method in the library.

While exploring slides on `RelevanceFeedback` and the Rocchio Algorithm, I am contemplating implementing the algorithm myself. This refiner holds significance and serves as the parent for other important refiners like RM3, BertQE, Termluster, and more.

All other refiners are functioning well, providing results for the gov2 dataset. During fixing issues, I encountered a minor issue with the `Anchor` and `Wiki` refiners. They faced challenges in calling and using their parent variables. Additionally, the recent version of `gensim `(4.x) removed the `vocab` attribute in the Word2Vec model, replacing it with `index_to_key`. I found a helpful resource [here](https://github.com/MilaNLProc/contextualized-topic-models/issues/114).

Currently, my focus is on resolving the issue with `RelevanceFeedback` along with working on RAG-fusion.

**Comment by DelaramRajaei** [comment]
Hello @hosseinfani,

I looked into a few more solutions to address the problem with the RelevanceFeedback refiner, but unfortunately, I couldn't find a successful fix. As a temporary measure, I'll stick to using only Anserini for this refiner until I come across a better solution. 

Here's the code snippet that utilizes Anserini:
```
    def get_tfidf(self, docid):
        # command = "target/appassembler/bin/IndexUtils -index lucene-index.robust04.pos+docvectors+rawdocs -dumpDocVector FBIS4-40260 -docVectorWeight TF_IDF "
        cli_cmd = f'\"./src/anserini/target/appassembler/bin/IndexUtils\" -index \"{self.index}\" -dumpDocVector \"{docid}\" -docVectorWeight TF_IDF'
        stream = os.popen(cli_cmd)
        return stream.read()
```

Meantime, I discovered some resources that might be useful in resolving the issue.

**Anerini**
[Extraction of TF-IDF vectors](https://github.com/castorini/anserini/blob/master/docs/runbook-trec2018-h2oloo.md)

**Pyserini**
[Pyserini: Reproducing Vector PRF Results](https://github.com/hosseinfani/ReQue/blob/master/qe/expanders/relevancefeedback.py)
[To Interpolate or not to Interpolate: PRF, Dense and Sparse Retrievers](https://cs.uwaterloo.ca/~jimmylin/publications/Li_etal_SIGIR2022.pdf)
[Pseudo Relevance Feedback with Deep Language Models and Dense Retrievers: Successes and Pitfalls](https://arxiv.org/pdf/2108.11044.pdf)
[Pseudo-Relevance Feedback with Dense Retrievers in Pyserini](https://dl.acm.org/doi/pdf/10.1145/3572960.3572982)

**Keywords Extraction Using TF-IDF Method**
[sklearn.feature_extraction.text](https://www.kdnuggets.com/2022/09/convert-text-documents-tfidf-matrix-tfidfvectorizer.html)
[sklearn.feature_extraction.text](https://kavita-ganesan.com/python-keyword-extraction/)
[gensim.models.TfidfModel](https://radimrehurek.com/gensim/models/tfidfmodel.html)

I came across this tool called Spacerini ([link](https://huggingface.co/papers/2302.14534)), which combines features from Pyserini and the Hugging Face ecosystem. It provides a simple and user-friendly method for researchers to explore and analyze large text datasets through interactive search applications. I'm not certain if we'll use it, but it could be helpful down the line.

**Comment by hosseinfani** [comment]
@DelaramRajaei 
thanks for the update. that's fine for the time being but create an issue page for it as a bug/issue so we can fix it in future. 

for code reference, you can paste the codeline permanent link at github like this:
https://github.com/fani-lab/RePair/blob/b752d8ecf7712b2c1134e6ac7f8ad9877f59ed6e/src/refinement/refiners/relevancefeedback.py#L56

**Comment by DelaramRajaei** [comment]
Hello @hosseinfani,

I've fixed the issues with RM3 and BertQ. Here's a brief overview of the changes:

**RM3:**
I noticed that RM3 in pyserini was only used for document reranking, and a similar approach was used to select the top word in the `Relevance feedback`. To address this, I updated the `get_topn_relevant_docids` function. The refiner now calls the `get_refined_query` from its parent, the `Relevance feedback`.

**BertQ:**
Dealing with BertQ was challenging due to its reliance on pygaggle for importing transformers, causing conflicts with other libraries. After reviewing their [paper](https://arxiv.org/abs/2009.07258#:~:text=BERT%2DQE%3A%20Contextualized%20Query%20Expansion%20for%20Document%20Re%2Dranking,-Zhi%20Zheng%2C%20Kai&text=Query%20expansion%20aims%20to%20mitigate,information%20when%20expanding%20the%20query.), I referred to this [link](https://www.linkedin.com/pulse/bert-ranking-yeshwanth-n/) and the [bert documentation](https://huggingface.co/docs/transformers/model_doc/bert#transformers.BertForSequenceClassification) for implementing the code as per their guidelines.

Both refiners are now working, and I've stored their results. While two other refiners (adoptonfields and onfields) are pending, my focus is currently on implementing Rag fusion and creating dense indexes to compare results with the existing refiners.

**Comment by DelaramRajaei** [comment]
Hello @hosseinfani,
I wanted to let you know about the work I've accomplished in the past weeks.

I've stored the outcomes of the refinement process applied to Rapir across all five datasets (robust04, gov2, antique, dbpedia, clueweb09b), where Sparse indices were available. Additionally, I've updated the Rapir's storyboard on Teams.

There have been changes to the pipeline, with the addition of more commands:

1. [`query_refinement`](https://github.com/DelaramRajaei/RePair/blob/7f257fc80891fd9979f2349daf7aa00af4bd9505/src/main.py#L56): This command triggers the execution of selected refiners in the [refiner.param](https://github.com/DelaramRajaei/RePair/blob/10ff8bb966275d97d77e5fa03b29837a7653841c/src/refinement/refiner_factory.py), including the original query. If the files already exist, they will be read and stored in the list of Query class objects for efficiency. If no refiners are selected only original query refiner will be called.

2. [`similarity`](https://github.com/DelaramRajaei/RePair/blob/7f257fc80891fd9979f2349daf7aa00af4bd9505/src/main.py#L69): This command computes [rouge, bleu, and semsim](https://github.com/DelaramRajaei/RePair/blob/7f257fc80891fd9979f2349daf7aa00af4bd9505/src/evl/trecw.py#L18) for all refined queries along with the original query. All the results will be stored in similarity folder. The output will be structured as follows:
```
├── output
│   └── dataset_name
│       └── similarity
│       └── refined_queries_files
```

3. [`rag_fusion`](https://github.com/DelaramRajaei/RePair/blob/7f257fc80891fd9979f2349daf7aa00af4bd9505/src/main.py#L151): This command gathers the outcomes of selected ranker for either all the refiners or just backtranslation, then calculates reciprocal_rank_fusion ([RRF](https://github.com/DelaramRajaei/RePair/blob/e99edb7d7af22b2831308402e969ee2bf770fc8f/src/dal/ds.py#L148)). Promising initial results have been achieved, although I'm still refining this step.

Additionally, several minor updates have been made to the Rapir project:

- The search function has been updated, and `search_df` has been removed. This change was made because Lucene doesn't allow multiple processes to modify an index simultaneously, as discussed [here](https://stackoverflow.com/questions/1211690/what-is-the-best-way-to-allow-multiple-processess-to-modify-a-lucene).
- [`set_index`](https://github.com/DelaramRajaei/RePair/blob/e99edb7d7af22b2831308402e969ee2bf770fc8f/src/dal/ds.py#L95) has been added to accommodate different index formats for msmarco and aol.
- The T5 model has been added as a refiner, and commands have been removed from the main pipeline. However, there have been some library conflicts between T5 and other refiners, leading to bugs during project execution.
- Efforts have been made to minimize redundancy in the code and to separate T5 settings from general settings.

Currently, my focus is on working on rag-fusion.
